components/calendar/calendar.js

```diff
@@ -0,0 +1,480 @@
+		if (date) {
+			if (this._keyboardTriggeredMonthChange) {
+				date.focus();
+				this._keyboardTriggeredMonthChange = false;
+			}
+		}
+	}
+
+	_getFocusDateElement() {
+		return this.shadowRoot.querySelector(`div[data-date="${this._focusDate.getDate()}"][data-month="${this._focusDate.getMonth()}"][data-year="${this._focusDate.getFullYear()}"]`);
+	}
+
+	_onDateSelected(e) {
+		const selectedDate = e.composedPath()[0];
+		const year = selectedDate.getAttribute('data-year');
+		const month = selectedDate.getAttribute('data-month');
+		const date = selectedDate.getAttribute('data-date');
+		this._focusDate = new Date(year, month, date);
+
+		this.selectedValue = (new Date(year, month, date)).toISOString();
+
+		const eventDetails = {
+			bubbles: true,
+			composed: true,
+			detail: { date: this.selectedValue }
+		};
+		this.dispatchEvent(new CustomEvent('d2l-calendar-selected', eventDetails));
+	}
+
+	_onKeyDown(e) {
+		const rootTarget = e.composedPath()[0];
+		if (!rootTarget.classList.contains('d2l-calendar-date')) return;
+
+		let preventDefault = false;
+		let numDaysChange;
+
+		switch (e.keyCode) {
+			case keyCodes.ENTER:
+			case keyCodes.SPACE:
+				preventDefault = true;
+				this._onDateSelected(e);
+				break;
+			case keyCodes.DOWN:
+				numDaysChange = daysInWeek;
+				preventDefault = true;
+				break;
+			case keyCodes.UP:
+				numDaysChange = -daysInWeek;
+				preventDefault = true;
+				break;
+			case keyCodes.LEFT:
+				if (getComputedStyle(this).direction === 'rtl') {
+					numDaysChange = 1;
+				} else {
+					numDaysChange = -1;
+				}
+				break;
+			case keyCodes.RIGHT:
+				if (getComputedStyle(this).direction === 'rtl') {
+					numDaysChange = -1;
+				} else {
+					numDaysChange = 1;
+				}
+				break;
+			case keyCodes.HOME: {
+				const dayOfTheWeek = this._focusDate.getDay();
+				if (getComputedStyle(this).direction === 'rtl') {
+					numDaysChange = 6 - dayOfTheWeek + firstDayOfWeek;
+					if (numDaysChange > 6) {
+						numDaysChange -= daysInWeek;
+					}
+				} else {
+					numDaysChange = dayOfTheWeek - firstDayOfWeek;
+					if (numDaysChange < 0) {
+						numDaysChange += daysInWeek;
+					}
+					numDaysChange *= -1;
+				}
+				preventDefault = true;
+				break;
+			} case keyCodes.END: {
+				const dayOfTheWeek = this._focusDate.getDay();
+				if (getComputedStyle(this).direction === 'rtl') {
+					numDaysChange = dayOfTheWeek - firstDayOfWeek;
+					if (numDaysChange < 0) {
+						numDaysChange += daysInWeek;
+					}
+					numDaysChange *= -1;
+				} else {
+					numDaysChange = 6 - dayOfTheWeek + firstDayOfWeek;
+					if (numDaysChange > 6) {
+						numDaysChange -= daysInWeek;
+					}
+				}
+				preventDefault = true;
+				break;
+			} case keyCodes.PAGEUP: {
+				if (e.shiftKey) {
+					// Changes the grid of dates to the previous Year.
+					// Sets focus on the same day of the same week. If that day does not exist, then moves focus to the same day of the previous or next week.
```

Comment:
Is the page up/down stuff TODO?

Reply:
It is, though it's in the design as a nice to have so it might be better to leave it out of the code completely and then add it if/when it's getting implemented.